### ISSUE TITLE
Road to v2

### ISSUE THREAD

**Issue by ctrlplusb** [ctrlplusb/code-split-component#8]
An alpha is up and running within `react-universally`.  It is tons better, but I would like us to finish all of the below before releasing an official v2.

- [ ] Test on webpack v1 and v2
- [x] Guard that the plugins are disabled for a module.hot environment. 
- [ ] Add example(s)
- [ ] Higher test coverage
- [x] API refactor/simplification parses
- [x] Clean up / document code.
- [x] Documentation
- [ ] A long standing issue-free alpha/beta version.

Nice to haves.

- [ ] Add optimised CodeSplitMatch component for react-router 4
- [x] react-router 3 story

**Comment by codepunkt** [comment]
`CodeSplitMatch` component would be awesome 👍

**Comment by codepunkt** [comment]
@ctrlplusb what kind of API did you have in mind for `CodeSplitMatch`? i'd love to help :)

Shortest i could come up with is this, but it isn't too flexible and leaves a lot to be desired (has to auto generate unique chunkName based on pattern, only a single module etc)
```javascript
<CodeSplitMatch pattern="/about" module={require('./About')} />
```

**Comment by ctrlplusb** [comment]
Help would be massively welcomed :)

I'll review the current `Match` API and come back with some ideas... _hopefully_. 😀

**Comment by codepunkt** [comment]
Questions that could be interesting in this regard:

- Is there any use case where multiple modules would be beneficial? If yes, which one?
- Is the `chunkName` used for anything else except as an internal unique identifier?

**Comment by swernerx** [comment]
`chunkName` is mainly used for file naming aka chunk names in webpack I think.

**Comment by ctrlplusb** [comment]
> Is there any use case where multiple modules would be beneficial? If yes, which one?

I actually struggle to think of one, because there is no harm in importing an individual module that could then import whatever other modules it needed.  Thinking about this I think we should simplify the `CodeSplit` API too.  Breaking change, but it is simpler.

> Is the chunkName used for anything else except as an internal unique identifier?

Actually, it's more helpful use is to group code split instances into a single chunk.  i.e. you can use the name in multiple places to indicate you would like all the respective modules within a chunk by the same name.  We could look into making this an optional value though and let our babel and webpack transpilers take care of the rest.

**Comment by ctrlplusb** [comment]
@codepunkt @swernerx - you guys aren't related are you? Popular surname? :D

**Comment by dlebedynskyi** [comment]
@ctrlplusb react-router is going throw some rewrite around Match and MatchRoutes. We are getting some issues around those with animations. Anyway alpha.6 is different that current master, so I'd be  careful about relying on react-router Match atm.

**Comment by swernerx** [comment]
@ctrlplusb I don't think so :)

**Comment by swernerx** [comment]
Mhh, what react-router version we are talking about? I mean the v4 has not a very steady development right now as it seems. https://github.com/ReactTraining/react-router/commits/v4

**Comment by ctrlplusb** [comment]
@dlebedynskyi do you have any references to these issues?

**Comment by dlebedynskyi** [comment]
rr4 is generally stable lib. we have been using it since alpha.4 And I really love declarative routing. But there are changes to context and createServerRenderContext(_like it is gone_) in current version4 branch. 

Animation related changes I mentioned https://github.com/ReactTraining/react-router/issues/4259#issuecomment-266069575. 
Note: version4 branch currently has different implementation that one in comment.

Also 26 days ago there was a release with comment
> has a heavily refactored version coming in the next day or two. 

There is version4-alt branch as well. 

tl;dr; There will be new MatchRoutes that will be rendering only one element. As well context changes are coming. Just something to pay attention too.

**Comment by ctrlplusb** [comment]
Project has been deprecated. 💖
